File: src/python/utils/time_utils.py
import re
from typing import int

def parse_duration(duration: str) -> int:
    """Parse duration string into seconds
    
    Args:
        duration: Duration string like '7d', '24h', '30m'
        
    Returns:
        Duration in seconds
        
    Raises:
        ValueError: If duration format is invalid
    """
    pattern = r'^(\d+)([hdwm])$'
    match = re.match(pattern, duration.lower())
    
    if not match:
        raise ValueError(f'Invalid duration format: {duration}. Use: 1h, 7d, 1w, etc.')
    
    value, unit = match.groups()
    value = int(value)
    
    multipliers = {
        'h': 3600,      # hours
        'd': 86400,     # days  
        'w': 604800,    # weeks
        'm': 2592000    # months (30 days)
    }
    
    return value * multipliers[unit]

def seconds_to_minutes(seconds: int) -> float:
    """Convert seconds to minutes"""
    return seconds / 60

def minutes_to_hours(minutes: float) -> float:
    """Convert minutes to hours"""
    return minutes / 60

File: src/js/utils/timeUtils.js
export function parseDuration(duration) {
  const match = duration.match(/^(\d+)([hm])$/);
  if (!match) {
    throw new Error('Invalid duration format. Use: 1h, 30m, etc.');
  }
  
  const [, value, unit] = match;
  const minutes = unit === 'h' ? parseInt(value) * 60 : parseInt(value);
  
  return minutes;
}

export function formatDuration(minutes) {
  if (minutes < 60) {
    return `${minutes}m`;
  } else {
    const hours = Math.floor(minutes / 60);
    const remainingMinutes = minutes % 60;
    return remainingMinutes > 0 ? `${hours}h${remainingMinutes}m` : `${hours}h`;
  }
}

export function formatTimestamp(timestamp) {
  return new Date(timestamp).toISOString();
}

export function hoursToMinutes(hours) {
  return hours * 60;
}

export function minutesToSeconds(minutes) {
  return minutes * 60;
}

export function calculateTimeSteps(durationMinutes, stepMinutes = 1) {
  return Math.ceil(durationMinutes / stepMinutes);
}